packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts

```diff
@@ -1414,5 +1518,93 @@ class Foo {
         },
       ],
     },
+    {
+      filename: 'test.ts',
+      code: `
+const foo = (function () {
+  return 'foo';
+})();
+      `,
+      options: [
+        {
+          allowIIFEs: false,
+        },
+      ],
+      errors: [
+        {
+          messageId: 'missingReturnType',
+          line: 2,
+          endLine: 2,
+          column: 14,
+          endColumn: 25,
+        },
+      ],
+    },
+    {
+      filename: 'test.ts',
+      code: `
+const foo = (function () {
+  return () => {
+    return 1;
+  };
+})();
+      `,
+      options: [
+        {
+          allowIIFEs: true,
+        },
+      ],
+      errors: [
+        {
+          messageId: 'missingReturnType',
+          line: 3,
+          endLine: 3,
+          column: 10,
+          endColumn: 15,
+        },
+      ],
+    },
+    {
+      filename: 'test.ts',
+      code: `
+let foo = function () {
+  return 'foo';
+};
+      `,
+      options: [
+        {
+          allowIIFEs: true,
+        },
+      ],
+      errors: [
+        {
+          messageId: 'missingReturnType',
+          line: 2,
+          endLine: 2,
+          column: 11,
+          endColumn: 22,
+        },
+      ],
+    },
+    {
+      filename: 'test.ts',
+      code: `
+let foo = (() => () => {})()();
```

Comment:
@JoshuaKGoldberg 
How should we handle multiple levels of IIFEs? For now, it allows only if the function is called directly.

```ts
const foo = (() => () => "foo")()(); // error

const bar = (() => (() => "foo")())(); // pass
```

Should both cases be passed when the `allowIIFEs` option is true?

Reply:
@yeonjuan The first one IMO is not an "IIFE" by definition—`(() => () => "foo")()` is an expression that happens to evaluate to a function, not a function expression itself. It's fine if we don't handle it.

Reply:
I would agree with that assessment.
The inner function is not an IIFE, just the outer function.